src/main/java/com/google/devtools/build/lib/analysis/SourceManifestAction.java

```diff
@@ -129,12 +138,48 @@ public SourceManifestAction(
       Artifact primaryOutput,
       Runfiles runfiles,
       boolean remotableSourceManifestActions) {
+    // The real set of inputs is computed in #getInputs().
     super(owner, NestedSetBuilder.emptySet(Order.STABLE_ORDER), primaryOutput, false);
     this.manifestWriter = manifestWriter;
     this.runfiles = runfiles;
     this.remotableSourceManifestActions = remotableSourceManifestActions;
   }
 
+  /**
+   * The manifest entry for a symlink artifact should contain the target of the symlink rather
+   * than its exec path. Reading the symlink target requires that the symlink artifact is declared
+   * as an input of this action. Since declaring all runfiles as inputs of the manifest action would
+   * unnecessarily delay its execution, this action exceptionally overrides
+   * {@link AbstractAction#getInputs()} and filters out the non-symlink runfiles by flattening the
+   * nested set of runfiles. Benchmarks confirmed that this does not regress performance.
+   *
+   * Alternatives considered:
+   * - Having users separate normal artifacts from symlink artifacts during analysis: Makes it
+   *   impossible to pass symlink artifacts to rules that aren't aware of them and requires the use
+   *   of custom providers to pass symlinks to stage as inputs to actions.
+   * - Reaching into {@link ActionExecutionContext} to look up the generating action of symlink
+   *   artifacts and retrieving the target from {@link UnresolvedSymlinkAction}: This would not work
+   *   for symlinks created by spawns.
+   * - Discovered inputs: ???
```

Comment:
Should I give input discovery a try or do you see a reason why it would perform worse?

Reply:
Incorporated the points explained in https://github.com/fmeum/bazel/commit/a2fdd4ad3963aec805d45ae1664a415bcdf2a3c3#r83921660.